### PR TITLE
[12.0][ADD] account_debitnote

### PR DESCRIPTION
**Pull request by kongrattapong**
In some situations, user need to add some service charge or some fees on top of what already invoiced.
This module add new feature to "Add Debit Note" from customer invoice / vendor bill

Ready to review.

### CODE REVIEW

File: account_debitnote/models/account_invoice.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    debit_invoice_id = fields.Many2one(
+        comodel_name='account.invoice',
+        string='Debit Note',
+        readonly=True,
+        states={'draft': [('readonly', False)]},
+    )
+    debit_invoice_ids = fields.One2many(
+        comodel_name='account.invoice',
+        inverse_name='debit_invoice_id',
+        string="Debit Notes",
```
Comment (kittiu) [inline account_debitnote/models/account_invoice.py]:
```suggestion
        string='Debit Notes',
```

File: account_debitnote/models/account_invoice.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+
+class AccountInvoice(models.Model):
+    _inherit = 'account.invoice'
+
+    debit_invoice_id = fields.Many2one(
+        comodel_name='account.invoice',
+        string='Debit Note',
+        readonly=True,
+        states={'draft': [('readonly', False)]},
```
Comment (kittiu) [inline account_debitnote/models/account_invoice.py]: Do we need to allow edtiable on draft?

File: account_debitnote/README.rst [hunk 3]
```diff
@@ -0,0 +1,88 @@
+===========
```
Comment (kittiu) [inline account_debitnote/README.rst:1]: You can remove this file. I will be create automatically once it is merged.

File: account_debitnote/models/account_invoice.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        comodel_name='account.invoice',
+        string='Debit Note',
+        help="Reference to the origin invoice that create this debit note",
+    )
+    debit_invoice_ids = fields.One2many(
+        comodel_name='account.invoice',
+        inverse_name='debit_invoice_id',
+        string='Debit Notes',
+        readonly=True,
+    )
```
Comment (kittiu) [inline account_debitnote/models/account_invoice.py:22]:
```suggestion
        help="List all debit notes being created by this invoice",
    )
```

File: account_debitnote/models/account_invoice.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    def _prepare_debitnote(self, invoice, date_invoice=None,
+                           date=None, description=None, journal_id=None):
+        """
+            Prepare the dict of values to create the new debit note
+            from the invoice.
+            This method may be overridden to implement custom debit note
+            generation (making sure to call super() to establish a clean
+            extension chain).
+
+            :param integer invoice_id: id of the invoice to create debit note
```
Comment (kittiu) [inline account_debitnote/models/account_invoice.py]:
```suggestion
```

File: account_debitnote/wizard/account_invoice_debitnote.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        inv_obj = self.env['account.invoice']
+        context = dict(self._context or {})
+        xml_id = False
+        for form in self:
+            created_inv = []
+            date = False
+            description = False
+            for inv in inv_obj.browse(context.get('active_ids')):
+                if inv.state in ['draft', 'cancel']:
+                    raise ValidationError(_('''Cannot create debit note for
```
Comment (kittiu) [inline account_debitnote/wizard/account_invoice_debitnote.py]:
```suggestion
                    raise ValidationError(_("""Cannot create debit note for
```

**Comment by kittiu** [comment]
We should not allow create Debit Note from Debit Note. Please hide "Create Debit Note" button, from the debit note itself. Also check the same in code.
Debit Note should be create from Invoice only.

Steps:
- From an Incoie, INV-001, I can create Debit Note, i.e., DINV-001 -> OK
- From Debit Note, i.e., DINV-001, I can create Debit Note, i.e., DINV-002 -> Not OK.

**Comment by kittiu** [comment]
image

File: account_debitnote/README.rst (hunk 3 above)
Comment (kongrattapong) [reply account_debitnote/README.rst:1]: Okay, I'll remove it.

File: account_debitnote/models/account_invoice.py (hunk 4 above)
Comment (kongrattapong) [reply account_debitnote/models/account_invoice.py:22]: roger that.
I will input it in.

File: account_debitnote/models/account_invoice.py (hunk 5 above)
Comment (kongrattapong) [reply account_debitnote/models/account_invoice.py]: roger that.

**Comment by kongrattapong** [comment]
> We should not allow create Debit Note from Debit Note. Please hide "Create Debit Note" button, from the debit note itself. Also check the same in code.
> Debit Note should be create from Invoice only.
> 
> Steps:
> 
> * From an Incoie, INV-001, I can create Debit Note, i.e., DINV-001 -> OK
> * From Debit Note, i.e., DINV-001, I can create Debit Note, i.e., DINV-002 -> Not OK.

Okay,I'll fix it.

**Comment by kongrattapong** [comment]
Customer Invoices (THB) - Odoo - Google Chrome_020

File: account_debitnote/readme/CONFIGURE.rst [hunk 7]
```diff
@@ -0,0 +1,5 @@
+To configure this module, you need to:
+
+- settings -> Users edit access right(Show Full Accounting Features).
```
Comment (kittiu) [inline account_debitnote/readme/CONFIGURE.rst]:
```suggestion
- Make sure user has access to "Show Full Accounting Features"
```

Comment (kittiu) [reply account_debitnote/readme/CONFIGURE.rst]: Please review the configuration steps.

File: account_debitnote/readme/DESCRIPTION.rst [hunk 8]
```diff
@@ -0,0 +1,3 @@
+In some situations, You need to add some service charge or add some fees
```
Comment (kittiu) [inline account_debitnote/readme/DESCRIPTION.rst]:
```suggestion
In some situations, user need to add some service charge or some fees on top of what already invoiced.
This module add new feature to "Create Debit Note" from customer invoice / vendor bill
```

File: account_debitnote/readme/DESCRIPTION.rst [hunk 9]
```diff
@@ -0,0 +1,3 @@
...
+you need to create debit note but in Invoicing don't have this module,
```
Comment (kittiu) [inline account_debitnote/readme/DESCRIPTION.rst]:
```suggestion
```

File: account_debitnote/readme/DESCRIPTION.rst [hunk 10]
```diff
@@ -0,0 +1,3 @@
...
+so I am interested to add this module.
```
Comment (kittiu) [inline account_debitnote/readme/DESCRIPTION.rst]:
```suggestion
```

File: account_debitnote/readme/USAGE.rst [hunk 11]
```diff
@@ -0,0 +1,10 @@
+To use this module, you need to:
+
+- Go to Invoicing -> Customers -> Invoices.
+    - Create a new invoice and Validate.
```
Comment (kittiu) [inline account_debitnote/readme/USAGE.rst]:
```suggestion
    - Create a new invoice and validate.
```

File: account_debitnote/readme/USAGE.rst [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+    - Click on Add Debit Note.
```
Comment (kittiu) [inline account_debitnote/readme/USAGE.rst]:
```suggestion
    - Click on Add Debit Note and fill in reasons. You can also specify an account date.
```

File: account_debitnote/readme/USAGE.rst [hunk 13]
```diff
@@ -0,0 +1,10 @@
...
+    - Get Reason. (You can add Account Date, If not Default is the same as the invoice.)
+    - Click Add Debit Note, You will have a new one Invoice(Debit Note).
+    - Validate Debit note.
+
+    * You can also do the same thing in Vendors -> Bills
```
Comment (kittiu) [inline account_debitnote/readme/USAGE.rst]:
```suggestion
You can also do the same with Vendor Bill.
```

**Review by kittiu — APPROVED** [review]

**Comment by kittiu** [comment]
Functional and technical reviewed.
@kongrattapong please squash commit.

**Comment by kittiu** [comment]
@OCA/accounting-maintainers

**Comment by kittiu** [comment]
@OCA/accounting-maintainers

**Comment by kittiu** [comment]
cc @ps-tubtim can you help review this?

**Review by ps-tubtim — APPROVED** [review]
Functional and Code review: OK.

**Pull request merged**